src/OgAccess.php

```diff
@@ -0,0 +1,202 @@
+use Drupal\user\EntityOwnerInterface;
+
+class OgAccess {
+
+  /**
+   * Administer permission string.
+   *
+   * @var string
+   */
+  const ADMINISTER_GROUP_PERMISSION = 'administer group';
+
+  /**
+   * Determines whether a user has a given privilege.
+   *
+   * All permission checks in OG should go through this function. This
+   * way, we guarantee consistent behavior, and ensure that the superuser
+   * and group administrators can perform all actions.
+   *
+   * @param $group_type
+   *   The entity type of the group.
+   * @param $group_id
+   *   The entity ID of the group.
+   * @param string $operation
+   *   The entity operation being checked for.
+   * @param $account
+   *   (optional) The account to check. Defaults to the current user.
+   * @param $skip_alter
+   *   (optional) If TRUE then user access will not be sent to other modules
+   *   using drupal_alter(). This can be used by modules implementing
+   *   hook_og_user_access_alter() that still want to use og_user_access(), but
+   *   without causing a recursion. Defaults to FALSE.
+   * @param $ignore_admin
+   *   (optional) When TRUE the specific permission is checked, ignoring the
+   *   "administer group" permission if the user has it. When FALSE, a user
+   *   with "administer group" will be granted all permissions.
+   *   Defaults to FALSE.
+   *
+   * @return bool
+   *   TRUE or FALSE if the current user has the requested permission.
+   *   NULL, if the given group isn't a valid group.
+   */
+  public static function userAccess($group_type, $group_id, $operation, $account = NULL, $skip_alter = FALSE, $ignore_admin = FALSE) {
+    $perm = &drupal_static(__FUNCTION__, []);
+    // Mark the group ID and permissions that invoked an alter.
+    $perm_alter = &drupal_static(__FUNCTION__ . '_alter', []);
+
+    if (!Og::isGroup($group_type, $group_id)) {
+      // Not a group.
+      return NULL;
+    }
+
+
+    if (empty($account)) {
+      $account = clone \Drupal::currentUser()->getAccount();
+    }
+
+    $account_id = $account->id();
+
+    // User #1 has all privileges.
+    if ($account_id == 1) {
+      return TRUE;
+    }
+
+    // Administer group permission.
+    if (!$ignore_admin && $account->hasPermission(static::ADMINISTER_GROUP_PERMISSION)) {
+      return TRUE;
+    }
+
+    // Group manager has all privileges (if variable is TRUE).
+    if (!empty($account_id) && \Drupal::config('og.settings')->get('group_manager_full_access')) {
+      $group = entity_load($group_type, $group_id);
+
+      if (($group instanceof EntityOwnerInterface) && ($group->getOwnerId() == $account_id)) {
+        return TRUE;
+      }
+    }
+
+    $identifier = $group_type . ':' . $group_id;
+
+    // To reduce the number of SQL queries, we cache the user's permissions
+    // in a static variable.
+    if (!isset($perm[$identifier][$account_id])) {
+      $perms = array();
+
+      if ($roles = og_get_user_roles($group_type, $group_id, $account_id)) {
```

Comment:
Yes, definitely agree! We need to at least get some chunks in so we can then spin up some smaller PRs around that. Should we comment this out for now maybe with a @todo ?